### PR TITLE
Issue/payload encoding

### PR DESCRIPTION
**Pull request by prolic**
I have still this failing test:

```
1) ProophTest\EventStore\Pdo\Projection\MySqlEventStoreProjectorCustomTablesTest::it_can_query_from_stream_and_filter_with_metadata_matcher
Failed asserting that 50 matches expected 1.

/home/sasa/code/prooph/pdo-event-store/vendor/prooph/event-store/tests/Projection/AbstractEventStoreProjectorTest.php:175
```

can you look at it?

### CODE REVIEW

**Comment by prolic** [comment]
ping @basz

**Comment by basz** [comment]
I'm not seeing that failure (after a rebase);

however; this is a (big) problem for MySQL.

```
DROP TABLE IF EXISTS test_json ;
CREATE TABLE `test_json` (
    `payload` JSON NOT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;

INSERT INTO test_json (payload) VALUES ('{"a float":-1.0}');
INSERT INTO test_json (payload) VALUES (JSON_OBJECT('a float', -1.0));
```
results
```
{"a float": -1}
{"a float": -1.0}
```

Seems like a bug with their string to json parsing.

**Comment by prolic** [comment]
Can we use `JSON_OBJECT` for inserting then?

Sascha-Oliver Prolic


Am So., 11. Nov. 2018 um 22:48 Uhr schrieb Bas Kamer <
notifications@github.com>:

**Comment by basz** [comment]
I think we could:

```
INSERT INTO test_json (payload) VALUES (JSON_OBJECT('an object', JSON_OBJECT('a float', -1.0, 'array', JSON_ARRAY(1.0, 2.0))));

```
```
{"an object": {"array": [1.0, 2.0], "a float": -1.0}}
```

**Comment by prolic** [comment]
But that means we would have to analyse the json structure upfront to
generate this code. Damn, what a mess.

Sascha-Oliver Prolic


Am So., 11. Nov. 2018 um 23:03 Uhr schrieb Bas Kamer <
notifications@github.com>:

**Comment by basz** [comment]
> what a mess.

yeah

**Comment by basz** [comment]
This is a brain-twister. 

The PersistenceStrategy::prepareData can flatten the payload simple enough, like so;

```
$data = array_merge($data, $this->flattenArray($eventData['payload']));
```
with 
```
private function flattenArray(array $payload): array
    {
        $data = [];
        foreach ($payload as $name => $value) {
            $data[] = $name;

            if (is_array($value)) {
                $data = array_merge($data, $this->flattenArray($value));
            } else {
                $data[] = $value;
            }
        }

        return $data;
    }
```

However, this covers the values. Now the MySqlEventStore needs to create a nested string with `JSON_OBJECT(?, JSON_OBJECT(?, ?), ?, ?)` and while we don't have the data structure anymore.

So I'm thinking let's do the analyses *in full* inside the strategy. 

```
    public function prepareData(Iterator $streamEvents): array
    {
        $data = [];
        $rowsSql = [];
        foreach ($streamEvents as $event) {
            $rowSql = [];
            $eventData = $this->messageConverter->convertToArray($event);

            if (! isset($eventData['metadata']['_aggregate_version'])) {
                throw new Exception\RuntimeException('_aggregate_version is missing in metadata');
            }

            $data[] = $eventData['metadata']['_aggregate_version'];
            $rowSql[] = '?';

            $data[] = $eventData['uuid'];
            $rowSql[] = '?';

            $data[] = $eventData['message_name'];
            $rowSql[] = '?';

            $data = array_merge($data, $this->flattenValues($eventData['payload']));
            $rowSql[] = $this->flattenPositions($eventData['payload']);
            
            $data[] = Json::encode($eventData['metadata']);
            $rowSql[] = '?';
            
            $data[] = $eventData['created_at']->format('Y-m-d\TH:i:s.u');
            $rowSql[] = '?';

            $rowsSql[] = $rowSql;
        }

        return $data;
    }

    public function generateTableName(StreamName $streamName): string
    {
        return '_' . \sha1($streamName->toString());
    }

    private function flattenValues(array $payload): array
    {
        $data = [];
        foreach ($payload as $name => $value) {
            $data[] = $name;

            if (is_array($value)) {
                $data = array_merge($data, $this->flattenValues($value));
            } else {
                $data[] = $value;
            }
        }

        return $data;
    }

    private function flattenPositions(array $payload): string
    {
        $string = 'JSON_OBJECT(';
        $positions = [];
        foreach ($payload as $name => $value) {
            $positions [] = '?';

            if (is_array($value)) {
                $positions [] = $this->flattenPositions($value);
            } else {
                $positions [] = '?';
            }
        }

        return $string . implode(',', $positions ). ')';
    }
```

`$rowsSql` becomes something like this. 

```(
    [0] => Array
        (
            [0] => ?
            [1] => ?
            [2] => JSON_OBJECT(?,JSON_OBJECT(?,?,?,?))
            [3] => ?
            [4] => ?
        )

)
```

And I think the position of values match. So we now need to expose this rowsSql to the event store.

I think I would solve this by returning both the prepared values and the prepared sql from the `prepareData` method.

I would, therefore, move this logic inside the EventStores related to rowsSql to the strategies.

```
        $rowPlaces = '(' . \implode(', ', \array_fill(0, \count($columnNames), '?')) . ')';
        $allPlaces = \implode(', ', \array_fill(0, $countEntries, $rowPlaces));
```

Thoughts?

I haven't given much thought to:
- do we still need to JSON::encode the values? I dont think so...
- Is JSON_OBJECT enough? What about JSON_ARRAY's? Inside the event payload there are only arrays or scalars. array's may be numerical indexed... What if one need to store `['a', 'b']`?

**Comment by prolic** [comment]
Maybe (because way easier): We don't add this `\JSON_PRESERVE_ZERO_FRACTION`
and put into the documentation that you have to handle this limitation
manually.

Sascha-Oliver Prolic


Am Mo., 12. Nov. 2018 um 18:52 Uhr schrieb Bas Kamer <
notifications@github.com>:

**Comment by basz** [comment]
That’s a way out, yes. However i feel that we should attempt to ensure that what goes in comes out exactly equally. In the end I don’t really care as I switched to Postgres for the event store a long time ago...

So before we give up; would you happen to know of a better way of doing this? Do you think I am onto the right track here?

**Comment by basz** [comment]
Given this more thought.

Apparently, JSON only knows a 'number' and implementations can do what the heck they want. Also in the context of an events-store, it may be safe (or advisable at the least) to assume value objects are used which should be able to cast values to whatever type is expected. Ánd the fact that we would get some yucky code for MySql that will perform less I now feel much more ambiguous to pursue this goal. It now feels like a struggle with no actual gains as people need to be aware of this limitation anyway.

However, mayby we should keep the `JSON_PRESERVE_ZERO_FRACTION` flag as that is what should happen in my opinion. Maybe we could mark the failing tests as warnings only for MySql? And some documentation, indeed.

tuesday morning, hands on quiz! type in your JS console;
```
JSON.parse(JSON.stringify(-1.0));
``` 
simply try, can you guess?
```
-1.0
```

**Comment by prolic** [comment]
@codeliner @enumag @fritz-gerneth @ocramius any thoughts?

**Comment by Ocramius** [comment]
Did anybody try to escalate it to MySQL itself first?

**Comment by fritz-gerneth** [comment]
I guess this is related to this issue https://bugs.mysql.com/bug.php?id=88230 . 

Agree with @basz to add this flag and point to that issue in MySql should someone want to rely to this behavior. This can also be solved in user-code by explicit casting to float when restoring your data from the payload too.

File: src/Util/Json.php [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        $flags = \JSON_UNESCAPED_UNICODE | \JSON_UNESCAPED_SLASHES | \JSON_PRESERVE_ZERO_FRACTION;
+
+        $string = \json_encode($value, $flags);
+
+        if ($error = \json_last_error()) {
+            throw new JsonException(\json_last_error_msg(), $error);
+        }
+
+        if ('[]' === $string) {
+            $string = '{}';
```
Comment (enumag) [inline src/Util/Json.php]: Why? Isn't `[]` valid json?

**Comment by enumag** [comment]
Discussion is TL;DR. Anything important there?

Code looks good except for the `[]` condition.

**Comment by prolic** [comment]
Empty json is {} in all other languages.

**Comment by prolic** [comment]
@basz same problem on MariaDB?

**Comment by prolic** [comment]
btw, it seems to be fixed in MySQL 8.0.4 (did not test myself). Maybe I make this the minimum requirement for new event-store v8 then :)

**Comment by enumag** [comment]
@prolic I don't know about other languages but in JavaScript `JSON.stringify([])` returns `'[]'`, not `'{}'`.

**Comment by prolic** [comment]
`[]` is totally valid JSON. The thing is, usually in all OOP languages you would constructor a JSON string from an object, hence an empty object in converted into `{}` (empty dictionary, not empty list). In `eventstore.org` f.e., when you query for stream metadata (which is a dictionary, not a list, hence represented as object, not as array) and there is nothing, you'll receive `{}` and not `[]`.

**Comment by fritz-gerneth** [comment]
> btw, it seems to be fixed in MySQL 8.0.4 (did not test myself). Maybe I make this the minimum requirement for new event-store v8 then :)

Since 5.7 will be supported for another two years / 8 has only reached GA 6 months ago I think we should give it a bit more time until we set it as minimum requirement. Especially if it is only due to such minor inconvinience.

**Comment by enumag** [comment]
I don't understand your point. Both `[]` and `{}` are valid JSONs. But they have different meaning - one is array, one is object.

**Comment by fritz-gerneth** [comment]
I'm with @enumag on this, for the same reason. Might this be a BC break for existing payloads with empty arrays as values? e.g. what does this return?

```
JSON::decode(JSON::encode([]));
```

**Comment by prolic** [comment]
php > echo json_encode(['foo' => 'bar']);
{"foo":"bar"}
php > echo json_encode([]);
[]

One is an array, the other one is also an array in PHP, but one is an object, the other one is an array in JSON. Event metadata, also probably also event data (when stored as JSON) will always be a JSON object (I can't really think of an event payload where you have an array), except when it's empty, then it should be an array?

**Comment by prolic** [comment]
Okay, I will check this again on event-store-client, maybe I can also adjust some tests and check for `{}` manually instead of relying on JSON::decode (which is kind of weird), but okay.

**Comment by enumag** [comment]
If your array is associative typecast it to `(object)` before calling `json_encode`.

php > echo json_encode((object) ['foo' => 'bar']);
{"foo":"bar"}
php > echo json_encode((object) []);
{}

**Comment by prolic** [comment]
Yes, maybe that's a good solution, let me double check this weekend.

Sascha-Oliver Prolic


Am Mi., 14. Nov. 2018 um 03:08 Uhr schrieb Jáchym Toušek <
notifications@github.com>:

**Comment by enumag** [comment]
Or better yet don't use associative arrays but stdClass. Both your hack and my typecast only solve this for root associative array but not for any nested associative array.

**Comment by prolic** [comment]
ok, changed.

**Review by enumag — APPROVED** [review]

**Pull request closed without merging**